File: 在线仓库/modelzoo-master/samples/common/infer/post_process/codeformer_postprocess.cpp
/*
 * Copyright (c) ModelZoo. 2025-2025. All rights reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "codeformer_postprocess.h"
#include "log.h"
#include "utils.h"
#include <opencv2/opencv.hpp>
#include <fstream>
#include <vector>
#include <string>
#include <sys/stat.h>
#include <algorithm>

namespace Infer {
namespace CodeFormerNS {

constexpr int TARGET_WIDTH = 512;
constexpr int TARGET_HEIGHT = 512;
constexpr int CHANNEL_NUM = 3;
constexpr float PIXEL_NORMALIZE_FACTOR = 255.0f;
constexpr float CLAMP_MIN = -1.0f;
constexpr float CLAMP_MAX = 1.0f;
constexpr float DENORM_OFFSET = 1.0f;
constexpr float DENORM_SCALE = 2.0f;
constexpr float CONVERT_OFFSET = 0.5f;

std::string GetFileName(const std::string& path) {
    size_t sep = path.find_last_of("/\\");
    size_t start = (sep == std::string::npos) ? 0 : sep + 1;
    size_t end = path.find_last_of(".");
    return (end == std::string::npos || end <= start) ?
           path.substr(start) : path.substr(start, end - start);
}

cv::Mat Postprocess(const float* preds, int originalHeight, int originalWidth) {
    // 转换为CHW格式Mat (3, H, W)
    cv::Mat chwMat(CHANNEL_NUM, TARGET_HEIGHT * TARGET_WIDTH, CV_32F);
    memcpy(chwMat.data, preds, CHANNEL_NUM * TARGET_HEIGHT * TARGET_WIDTH * sizeof(float));

    // 裁剪到[-1, 1]并映射到[0, 1]
    for (size_t i = 0; i < chwMat.total(); ++i) {
        float val = chwMat.at<float>(i);
        val = std::max(CLAMP_MIN, std::min(CLAMP_MAX, val));
        chwMat.at<float>(i) = (val + DENORM_OFFSET) / DENORM_SCALE;
    }

    // CHW转HWC
    std::vector<cv::Mat> channels(CHANNEL_NUM);
    for (int c = 0; c < CHANNEL_NUM; ++c) {
        channels[c] = cv::Mat(TARGET_HEIGHT, TARGET_WIDTH, CV_32F,
                             chwMat.ptr<float>(c)).clone();
    }
    cv::Mat hwcMat;
    cv::merge(channels, hwcMat);

    // RGB转BGR
    cv::Mat bgrMat;
    cv::cvtColor(hwcMat, bgrMat, cv::COLOR_RGB2BGR);

    // 转换为uint8
    cv::Mat uint8Mat;
    bgrMat.convertTo(uint8Mat, CV_8UC3, PIXEL_NORMALIZE_FACTOR, CONVERT_OFFSET);

    // 调整回原始尺寸
    if (originalHeight != TARGET_HEIGHT || originalWidth != TARGET_WIDTH) {
        cv::resize(uint8Mat, uint8Mat, cv::Size(originalWidth, originalHeight),
                  0, 0, cv::INTER_LINEAR);
    }

    return uint8Mat;
}

bool CodeFormerPostprocess(std::vector<std::string>& fileList,
                          std::vector<TensorBuf>& outBufs,
                          std::vector<TensorDesc>& outDescs) {
    LOG(INFO) << "====== 开始后处理 ======";
    auto cfg = ReadCfgFile("../data/cfg.txt");
    std::string saveBin = cfg["save_result_bin"];
    std::string saveImg = cfg["save_result_img"];

    if (!saveBin.empty() && !CreateDirectoryRecursive(saveBin)) {
        LOG(ERROR) << "创建bin目录失败: " << saveBin;
        return false;
    }
    if (!saveImg.empty() && !CreateDirectoryRecursive(saveImg)) {
        LOG(ERROR) << "创建图像目录失败: " << saveImg;
        return false;
    }

    for (size_t imgIdx = 0; imgIdx < fileList.size(); ++imgIdx) {
        const std::string& imgPath = fileList[imgIdx];
        std::string fileName = GetFileName(imgPath);
        LOG(INFO) << "处理文件: " << fileName;

        // 保存原始输出bin
        if (!saveBin.empty()) {
            std::string binPath = saveBin + "/" + fileName + "_output_0.bin";
            std::ofstream binFile(binPath, std::ios::binary);
            binFile.write(reinterpret_cast<char*>(outBufs[imgIdx].data.get()),
                         outBufs[imgIdx].size);
            LOG(INFO) << "保存原始输出bin to: " << binPath;
        }

        // 验证输出数据
        const float* preds = reinterpret_cast<const float*>(outBufs[imgIdx].data.get());
        size_t expectedSize = CHANNEL_NUM * TARGET_HEIGHT * TARGET_WIDTH * sizeof(float);
        if (outBufs[imgIdx].size != expectedSize) {
            LOG(ERROR) << "输出尺寸不匹配，预期: " << expectedSize
                      << "，实际: " << outBufs[imgIdx].size;
            continue;
        }

        // 获取原始图像尺寸
        cv::Mat originalImg = cv::imread(imgPath);
        if (originalImg.empty()) {
            LOG(WARNING) << "无法读取原始图像: " << imgPath;
            continue;
        }
        int originalHeight = originalImg.rows;
        int originalWidth = originalImg.cols;

        // 执行后处理
        cv::Mat resultImg = Postprocess(preds, originalHeight, originalWidth);

        // 保存结果图像
        if (!saveImg.empty()) {
            std::string imgSavePath = saveImg + "/" + fileName + "_restored.jpg";
            if (!cv::imwrite(imgSavePath, resultImg)) {
                LOG(WARNING) << "保存结果图像失败: " << imgSavePath;
            } else {
                LOG(INFO) << "已保存修复结果: " << imgSavePath;
            }
        }
    }

    LOG(INFO) << "====== 后处理完成 ======";
    return true;
}

}  // namespace CodeFormer
}  // namespace Infer

File: 在线仓库/modelzoo-master/samples/built-in/super_resolution/CodeFormer/script/codeformer_postprocess.py
# Copyright (c) ModelZoo. 2025-2025. All rights reserved.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import numpy as np
import cv2
import argparse


def postprocess(preds, original_h, original_w, target_h, target_w):
    """后处理：将模型输出转换为原始尺寸的图像"""
    # 处理模型输出
    img_np = preds[0].clip(-1, 1)  # 裁剪到[-1, 1]范围
    img_np = (img_np + 1) / 2  # 映射到[0, 1]

    # 转换为HWC格式并从RGB转BGR
    img_np = img_np.transpose(1, 2, 0)
    img_np = cv2.cvtColor(img_np, cv2.COLOR_RGB2BGR)

    # 转换为uint8格式
    img_np = (img_np * 255.0).round().astype(np.uint8)

    # 调整回原始图像尺寸
    if (img_np.shape[1], img_np.shape[0]) != (original_w, original_h):
        img_np = cv2.resize(img_np, (original_w, original_h), interpolation=cv2.INTER_LINEAR)

    return img_np


def run_postprocess(args):
    """执行后处理并保存结果图像"""
    os.makedirs(args.output_dir, exist_ok=True)

    # 获取所有推理结果文件
    bin_files = [f for f in os.listdir(args.bin_dir) if f.endswith('_output_0.bin')]
    if not bin_files:
        print(f"错误: 在 {args.bin_dir} 中未找到任何输出bin文件")
        return

    # 处理每个推理结果
    total = len(bin_files)
    target_h, target_w = args.target_size
    for idx, bin_file in enumerate(bin_files):
        # 提取图像名称
        img_name = bin_file.split("_output_0.bin")[0]
        img_path = os.path.join(args.img_dir, f"{img_name}.jpg")
        if not os.path.exists(img_path):
            print(f"警告: 原始图像 {img_path} 不存在，跳过")
            continue

        # 获取原始图像尺寸
        image = cv2.imread(img_path)
        if image is None:
            print(f"警告: 无法读取图像 {img_path}，跳过")
            continue
        original_h, original_w = image.shape[:2]

        # 加载推理结果
        bin_path = os.path.join(args.bin_dir, bin_file)
        try:
            preds = np.fromfile(bin_path, dtype=np.float32)
            # 恢复为模型输出形状 (1, 3, H, W)
            preds = preds.reshape(1, 3, target_h, target_w)
        except Exception as e:
            print(f"警告: 处理文件 {bin_file} 失败: {e}，跳过")
            continue

        # 后处理
        result_img = postprocess(
            preds,
            original_h=original_h,
            original_w=original_w,
            target_h=target_h,
            target_w=target_w
        )

        # 保存结果图像
        output_img_path = os.path.join(args.output_dir, f"{img_name}_restored.jpg")
        try:
            cv2.imwrite(output_img_path, result_img)
        except Exception as e:
            print(f"警告: 保存结果图像 {output_img_path} 失败: {e}，跳过")
            continue

        # 打印进度
        if (idx + 1) % 10 == 0 or (idx + 1) == total:
            progress = (idx + 1) / total * 100
            print(f"处理进度: {idx + 1}/{total} ({progress:.1f}%) - 文件: {bin_file}")

    print(f"所有后处理结果已保存至: {args.output_dir}")


def main():
    parser = argparse.ArgumentParser(description='CodeFormer后处理脚本')
    parser.add_argument('--bin_dir', default='../out/result_pc/bin',
                        help='存放推理输出bin文件的目录')
    parser.add_argument('--img_dir', default='../../../../../datasets/celeba-512/celeba-512-300/hq/',
                        help='原始图像目录')
    parser.add_argument('--output_dir', default='../out/result_pc/img',
                        help='输出结果图像目录')
    parser.add_argument('--target_size', type=int, nargs=2, default=[512, 512],
                        help='模型输入尺寸，格式为 高 宽，默认512 512')

    args = parser.parse_args()

    # 验证输入目录
    if not os.path.isdir(args.bin_dir):
        print(f"错误: 推理结果目录 {args.bin_dir} 不存在")
        return
    if not os.path.isdir(args.img_dir):
        print(f"错误: 原始图像目录 {args.img_dir} 不存在")
        return

    run_postprocess(args)


if __name__ == "__main__":
    main()